server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java

```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
 
   @Override
   protected boolean hasProjectUuidPermission(String permission, String projectUuid) {
-    if (!projectPermissionsCheckedByUuid.contains(permission)) {
-      try (DbSession dbSession = dbClient.openSession(false)) {
-        // FIXME do not load all the authorized projects. It can be huge.
-        Collection<String> projectUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, getUserId(), permission);
-        addProjectPermission(permission, projectUuids);
-      }
+    if (permissionsByProjectUuid == null) {
+      permissionsByProjectUuid = HashMultimap.create();
     }
-    return projectUuidByPermission.get(permission).contains(projectUuid);
+    Set<String> permissions;
+    if (permissionsByProjectUuid.containsKey(projectUuid)) {
+      permissions = permissionsByProjectUuid.get(projectUuid);
+    } else {
+      permissions = loadProjectPermissions(projectUuid);
+      permissionsByProjectUuid.putAll(projectUuid, permissions);
+    }
+    return permissions.contains(permission);
   }
 
-  private void addProjectPermission(String permission, Collection<String> authorizedProjectUuids) {
-    for (String key : authorizedProjectUuids) {
-      projectUuidByPermission.put(permission, key);
+  private Set<String> loadProjectPermissions(String projectUuid) {
+    Set<String> permissions;
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      if (userDto != null && userDto.getId() != null) {
+        permissions = dbClient.authorizationDao().selectProjectPermissions(dbSession, projectUuid, userDto.getId());
```

Comment:
I favor multiple returns over undefined variable at the beginning of method. So old-school procedural style... :p